### PR TITLE
perf: compute twiddle cache just once per degree

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR optimizes `Radix2EvaluationDomain` by introducing `TwiddleCache<F>`.

### CODE REVIEW

**Comment by TomTaehoonKim** [comment]
How about just dropping [3d3cbaf](https://github.com/kroma-network/tachyon/pull/530/commits/3d3cbaf2bdad7f757c2955a526867016956af458)?

File: tachyon/math/polynomials/univariate/radix2_twiddle_cache.h [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      for (size_t i = 1; i < domain->log_size_of_group(); ++i) {
+        size /= 2;
+        roots_vec[domain->log_size_of_group() - i - 1].resize(size);
+        inv_roots_vec[i].resize(size);
+      }
+
+      // Assign every element based on the biggest vector.
+      OMP_PARALLEL_FOR(size_t i = 1; i < domain->log_size_of_group(); ++i) {
+        for (size_t j = 0; j < domain->size() / std::pow(2, i + 1); ++j) {
+          size_t k = std::pow(2, i) * j;
```
Comment (GideokKim) [inline tachyon/math/polynomials/univariate/radix2_twiddle_cache.h:115]: How about this?

```suggestion
        size_t pow2_i = size_t{1} << i;
        for (size_t j = 0; j < domain->size() / (pow2_i << 1); ++j) {
          size_t k = pow2_i * j;
```

**Comment by chokobole** [comment]
> How about just dropping [3d3cbaf](https://github.com/kroma-network/tachyon/pull/530/commits/3d3cbaf2bdad7f757c2955a526867016956af458)?

It's possible since the change is minor, and most of people wouldn't care how the comment is modified. I can do that, but why do you suggest dropping?

**Review by GideokKim — APPROVED** [review]
LGTM

**Comment by TomTaehoonKim** [comment]
@chokobole I thought it was unnecessary but nvm.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/zk/plonk/permutation/unpermuted_table.h [hunk 2]
```diff
@@ -57,7 +57,7 @@ class UnpermutedTable {
                                    const Domain* domain) {
     // The ω is gᵀ with order 2ˢ where modulus = 2ˢ * T + 1.
     std::vector<F> omega_powers =
-        domain->GetRootsOfUnity(rows, domain->group_gen());
+        Domain::GetRootsOfUnity(rows, domain->group_gen());
```
Comment (batzor) [inline tachyon/zk/plonk/permutation/unpermuted_table.h:60]: Can you do the same for unittests as well?

File: tachyon/math/polynomials/univariate/radix2_twiddle_cache.h [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    //   [1, ω⁻¹, ω⁻², ω⁻³, ω⁻⁴, ω⁻⁵, ω⁻⁶, ω⁻⁷, ω⁻⁸, ω⁻⁹, ω⁻¹⁰, ω⁻¹¹, ω⁻¹², ω⁻¹³, ω⁻¹⁴, ω⁻¹⁵],
+    //   [1, ω⁻², ω⁻⁴, ω⁻⁶, ω⁻⁸, ω⁻¹⁰, ω⁻¹², ω⁻¹⁴],
+    //   [1, ω⁻⁴, ω⁻⁸, ω⁻¹²],
+    //   [1, ω⁻⁸],
+    //   [1],
+    // ]
+    // clang-format on
+    template <typename Domain>
+    CONSTEXPR_IF_NOT_OPENMP void Init(const Domain* domain,
+                                      bool packed_vec_only) {
```
Comment (batzor) [inline tachyon/math/polynomials/univariate/radix2_twiddle_cache.h:62]: Is there a reason why it's not a constructor?

**Comment by chokobole** [comment]
Closed this PR because the CI frequently failed to run, and the error is difficult to reproduce locally.

**Pull request closed without merging**